pkg/build/controller/build/build_controller.go

```diff
@@ -966,23 +966,54 @@ func (bc *BuildController) resolveImageStreamTag(ref *corev1.ObjectReference, li
 		}
 		return nil, fmt.Errorf("the referenced image stream %s/%s could not be found: %v", namespace, name, err)
 	}
-	if newRef, ok := imageutil.ResolveLatestTaggedImage(stream, tag); ok {
+	newRef := ""
+	ok = false
+	if len(stream.Status.DockerImageRepository) > 0 {
+		newRef, ok = imageutil.ResolveLatestTaggedImage(stream, tag)
+	} else {
+		streamCopy := stream.DeepCopy()
+		streamCopy.Status.DockerImageRepository = bc.internalRegistryHostname
+		newRef, ok = imageutil.ResolveLatestTaggedImage(streamCopy, tag)
+	}
+	if ok {
 		// generate informational event if
 		// a) tag has spec
 		// b) tag is local reference
-		// c) status DockerImageRepository is not set, meaning we got this copy of the stream from the watch
+		// c1) status DockerImageRepository is not set, meaning we got this copy of the stream from the watch
 		//    before the image registry was set up
+		// c2) or the import has not happened yet
 		// note, this method is only called for non-output image refs, so we should not get duplicates with
 		// the error/warning currently in place for the output link falling into this category
 		ref2, ok2 := imageutil.SpecHasTag(stream, tag)
-		if ok2 && ref2.ReferencePolicy.Type == imagev1.LocalTagReferencePolicy && len(stream.Status.DockerImageRepository) == 0 {
-			bc.recorder.Eventf(build, corev1.EventTypeNormal,
-				"LocalRefImageStreamTagUnavailable",
-				"The build %s/%s references an image stream tag %s with the local reference policy type, but the internal image registry location was not marked in the image stream",
-				build.Namespace,
-				build.Name,
-				ref.Name)
-		}
+		if ok2 && ref2.ReferencePolicy.Type == imagev1.LocalTagReferencePolicy {
+			if len(stream.Status.DockerImageRepository) == 0 {
+				bc.recorder.Eventf(build, corev1.EventTypeNormal,
+					"LocalRefImageStreamTagUnavailableNoImageRegistryRef",
+					"The build %s/%s references an image stream tag %s with the local reference policy type, but the internal image registry location was not marked in the image stream",
+					build.Namespace,
+					build.Name,
+					ref.Name)
+			}
+			latestTaggedImage := imageutil.LatestTaggedImage(stream, tag)
+			if len(latestTaggedImage.Image) == 0 {
+				//TODO len(latestTaggedImage.Image) == 0 can also result in not using the IST
+				// local tag reference policy;
+				// see https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L297
+				// see https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L304
+				// see https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L333
+				// if stream.Status.DockerImageRepository
+				// is empty, how likely is it that the Image field will be set?
+				// should we do something "crazy" like poll/wait on the stream??
```

Comment:
no.  it's valid to define an imagestream that uses localref policy, and not have an internal registry configured.

in such a case, i'd expect imagestream resolution to return the external registry spec.  Whether we can then pull that spec or not is unknown, but it's certainly *possible* if the user has the right creds in place.

Anyway that's the behavior we had implemented (if not local registry exists, use the external registry spec when resolving the imagestream) so i think we should preserve it.

Reply:
OK, as a reminder though, what is biting us in the extended tests is that the external registry spec is to the TBR, and our tests do not have the TBR creds included in the build's pull secret.

Pending the clarification on https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L333 , and if we can't further mitigate the "current imagestream entry in the lister cache is stale, etc.", perhaps we should start looking at a utility in the extended tests to mine the TBR creds from the install pull secret and seeding it as a pull secret in our extended test build pull secrets, just in case.

Reply:
> OK, as a reminder though, what is biting us in the extended tests is that the external registry spec is to the TBR, and our tests do not have the TBR creds included in the build's pull secret.

yes, understood, but the intent of your fix here is that we will properly resolve the imagestreamtag to the local reference because we are not relying on the internal registry hostname being set in the copy of the imagestream that we have, which might be stale, but instead are using the internal registry hostname value that the OCM knows about, which is not stale because the OCM gets restarted when that value becomes available.

The theoretical timing window here is:

1) internal registry hostname becomes available in the cluster
2) OCM restarts, caches imagestream values.  Because the apiserver has not restarted yet, the imagestream values the OCM gets, do not have the registry hostname decorated onto them.  The cache will never be "fixed" unless the imagestream changes again for some reason.
3) apiserver restarts to pick up internal registry hostname (decorator logic for gets is now in effect, but it's too late).
4) OCM starts a build, uses the cached imagestream that lacks the registry hostname to resolve a tag, so it resolves the tag to the external registry.

The intent of your changes is that you will not be reliant on the internal registry hostname being set on the imagestream via the decorator, which means you are not dependent on the apiserver restarting before OCM restarts.